packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+  return {
+    claims: cborDecode(hexToU8a(credential.claims.toHex())),
+    cTypeHash: credential.ctypeHash.toHex(),
+    delegationId: credential.authorization.unwrapOr(undefined)?.toHex() ?? null,
+    subject: credential.subject.toUtf8() as AssetDidUri,
+  }
+}
+
+// Retrieve a given block and looks into it to find a public credential creation tx that matches the provided credential ID.
+async function retrievePublicCredentialCreationExtrinsicsFromBlock(
+  api: ApiPromise,
+  credentialId: HexString,
+  blockNumber: u64
+): Promise<Extrinsic | null> {
+  const { extrinsics } = await api.derive.chain.getBlockByNumber(blockNumber)
+  return (
+    extrinsics
+      // Filter out failed extrinsics
+      .filter(({ dispatchError }) => dispatchError === undefined)
+      // Filter out extrinsics that don't contain a `CredentialStored` event with the right credential ID
+      .filter(({ events }) =>
+        events.some(
+          (event) =>
+            api.events.publicCredentials.CredentialStored.is(event) &&
+            event.data[1].toString() === credentialId
+        )
+      )
+      // If there is more than one (e.g., same credential issued multiple times in the same block) it should not matter since the ID is generated over the content, hence same ID -> same content.
+      // Nevertheless, take only the last one, if present, as that is for sure what ended up being in the blockchain state.
+      .pop()?.extrinsic ?? null
+  )
+}
+
+/**
+ * Decodes the public credential details returned by `api.call.publicCredential.getCredential()`.
+ *
+ * This is the **only** secure way for users to retrieve and verify a credential.
+ * Hence, calling `api.call.publicCredentials.getCredential(credentialId)` and then passing the result to this function is the only way to trust that a credential with a given ID is valid.
+ *
+ * @param credentialId Credential ID to use for the query. It is required to complement the information stored on the blockchain in a [[PublicCredentialsCredentialsCredentialEntry]].
+ * @param publicCredentialEntry The raw public credential details from blockchain.
+ * @returns The [[IPublicCredential]] as the result of combining the on-chain information and the information present in the tx history.
+ */
+export async function credentialFromChain(
+  credentialId: HexString,
+  publicCredentialEntry: Option<PublicCredentialsCredentialsCredentialEntry>
+): Promise<IPublicCredential> {
+  const api = ConfigService.get('api')
+
+  const { blockNumber, revoked } = publicCredentialEntry.unwrap()
+
+  const extrinsic = await retrievePublicCredentialCreationExtrinsicsFromBlock(
+    api,
+    credentialId,
+    blockNumber
+  )
+
+  if (extrinsic === null) {
+    throw new SDKErrors.PublicCredentialError(
+      `The block number as specified in the provided credential entry (${blockNumber}) does not have any extrinsic that includes a credential creation.`
+    )
+  }
+
+  if (!api.tx.did.submitDidCall.is(extrinsic)) {
```

Comment:
Do we need to do this check provided we have already checked `api.events.publicCredentials.CredentialStored.is(event)` above?

Reply:
Yes, because the event could be generated in multiple ways in the future, and this code makes sure that our assumptions are indeed correct. I think this is a viable approach, don't you think?

Reply:
Okay, I was just making sure :)